src/sinks/datadog/logs/log_api.rs

```diff
@@ -0,0 +1,308 @@
+}
+
+impl<Client> LogApi<Client>
+where
+    Client: Service<Request<Body>, Response = http::Response<Body>> + Send + Unpin,
+    Client::Future: Send,
+    Client::Response: Send + Debug,
+    Client::Error: Send + Debug,
+{
+    pub fn new() -> LogApiBuilder<Client> {
+        LogApiBuilder::default().bytes_stored_limit(bytesize::mib(5_u32))
+    }
+}
+
+#[async_trait]
+impl<Client> StreamSink for LogApi<Client>
+where
+    Client: Service<Request<Body>, Response = Response<Body>> + Send + Sync + Unpin,
+    Client::Future: Send,
+    Client::Error: Send + Debug,
+{
+    async fn run(&mut self, input: BoxStream<'_, Event>) -> Result<(), ()> {
+        // copy items out of `self`, needed as we don't have ownership
+        let compression = self.compression;
+        let datadog_uri = self.datadog_uri.clone();
+        let default_api_key = Arc::<str>::clone(&self.default_api_key);
+        let encoding = self.encoding.clone();
+        let host_key = self.log_schema_host_key;
+        let message_key = self.log_schema_message_key;
+        let timestamp_key = self.log_schema_timestamp_key;
+
+        let input = input.map(|mut event| {
+            let log = event.as_mut_log();
+            log.rename_key_flat(message_key, "message");
+            log.rename_key_flat(timestamp_key, "date");
+            log.rename_key_flat(host_key, "host");
+            encoding.apply_rules(&mut event);
+            event
+        });
+        let input =
+            Batcher::new(
+                input,
+                EventPartitioner::default(),
+                self.timeout,
+                NonZeroUsize::new(MAX_PAYLOAD_ARRAY).unwrap(),
+                None,
+            )
+            .map(|(api_key, events): (Option<Arc<str>>, Vec<Event>)| {
+                let api_key = api_key.unwrap_or_else(|| Arc::<str>::clone(&default_api_key));
+                let (fields, finalizers) = dissect_batch(events);
+                let uri = datadog_uri.clone();
+                let request = build_request(&fields, &api_key, uri, &compression);
+                (request, finalizers)
+            })
+            .filter_map(
+                |(request, finalizers): (
+                    Result<ApiRequest, FlushError>,
+                    Vec<EventFinalizers>,
+                )| async move {
+                    if request.is_err() {
+                        // TODO must log this error
+                        for finalizer in finalizers {
+                            finalizer.update_status(EventStatus::Failed);
+                        }
+                        return None;
+                    }
+                    Some((request.unwrap(), finalizers))
+                },
+            );
+        let mut input = Box::pin(input);
+
+        while let Some((api_request, finalizers)) = input.next().await {
+            flush_to_api(&mut self.http_client, api_request.inner, finalizers)
+                .expect("unhandled error")
+                .await;
```

Comment:
With this `await`, I'm not sure that we'll make any concurrent requests. It seems like we'd need some spawning and tracking of in-flight requests, similar to what we had in #8884.

Reply:
Agreed. I really wanted to avoid having to introduce a futuresunordered for the implementation to manually tend to, but I think that's a step too far right now.